server/bin/pbench-dispatch.sh

```diff
@@ -3,222 +3,284 @@
+    tbname=${resultname}.tar.xz
+    restorecon ${dest}/${tbname} ${dest}/${tbname}.md5
+    sts=${?}
+    if [[ ${sts} -ne 0 ]]; then
+        # log it but do not abort
+        log_error "${TS}: Error: \"restorecon ${dest}/${tbname} ${dest}/${tbname}.md5\", status ${sts}" "${status}"
+    fi
+
+    # Now that we have successfully moved the tar ball and its .md5 to the
+    # destination, we can remove the original .md5 file.
+    rm ${tb}.md5
+    sts=${?}
+    if [[ ${sts} -ne 0 ]]; then
+        log_error "${TS}: Warning: cleanup of successful copy operation failed: \"rm ${tb}.md5\", status ${sts}" "${status}"
+    fi
+
+    # Create a link in each state dir - if any fail we don't delete them
+    # because we have race conditions with other cron jobs.
     let toterr=0
     let totsuc=0
-    for state in $linkdestlist ;do
-        ln -sf $link ${controller_path}/${state}/
-        status=$?
-        if [ $status -eq 0 ] ;then
-            let totsuc+=1
+    for state in ${linkdestlist}; do
+        ln -sf ${dest}/${tbname} ${dest}/${state}/
+        sts=${?}
+        if [[ ${sts} -eq 0 ]]; then
+            (( totsuc++ ))
         else
-            log_error "$TS: Cannot create $tarball link to $state: code $status" "${mail_content}"
-            let toterr+=1
+            log_error "${TS}: Cannot create ${dest}/${tbname} link to ${state}: code ${sts}" "${status}"
+            (( toterr++ ))
         fi
     done
-    if [ $toterr -gt 0 ]; then
+    if [[ ${toterr} -gt 0 ]]; then
         # Count N link creations as one error since it is for handling of a
         # single tar ball.
-        let nerrs+=1
+        (( nerrs++ ))
     fi
-    if [ $totsuc -gt 0 ]; then
-        # We had at least one successful link state creation above.  So
-        # it is safe to remove the original link, as we'll use the logs
-        # to track down how to recover.
-        rm -f $tarball
-        status=$?
-        if [ $status -ne 0 ] ;then
-            log_error "$TS: Cannot remove $tarball link: code $status" "${mail_content}"
-            if [ $toterr -eq 0 ]; then
-                # We had other errors already counted against the total
-                # so don't bother counting this error
-                let nerrs+=1
-            fi
-        fi
-        let ntb+=1
-        if [ $toterr -gt 0 ]; then
-            # We have had some errors while processing this tar ball, so
-            # count this as a partial success.
-            log_info "$TS: $controller/$resultname: success (partial)"
-            let npartialsucc+=1
+    if [[ ${totsuc} -gt 0 ]]; then
+        # We had at least one successful link state creation above.
+        (( ntbs++ ))
+        if [[ ${toterr} -gt 0 ]]; then
+            # We have had some errors while processing this tar ball, so count
+            # this as a partial success.
+            log_info "${TS}: ${controller}/${resultname}: success (partial)" "${status}"
+            (( npartialsucc++ ))
```

Comment:
what will we do with partially successful tarballs?

Reply:
I don't think there's ever been any automated corrective procedure here. In theory I suppose it could remember the list of failures and retry: but I would think failures are unlikely, and a retry in those cases probably wouldn't succeed.

However: yeah, you're on to a good idea: collecting a *list* of the failed links and reporting them in the log would let someone create them manually later.

Then again, this PR focuses more on retaining the existing behavior with a restructure rather than improving things, so that's possibly out of scope.

Reply:
So the `log_info` message is in the log, so they are recorded.

The big thing is backup.  If a symlink here fails to create the `TO-BACKUP` it won't get backed up.

Reply:
And I agree, I don't think we want to change too much here, except the backup story, which we might want to make happen directly.

I think I'll propose something in another PR.